### DISCUSSION TITLE
sqflite _onCreate completes asynchronously

### DISCUSSION THREAD

**Discussion by myxiplx** [tekartik/sqflite#582] [General]
I'm having a problem with the way the database helpers initialize, create and update the database.  Since these functions are making async calls to _onCreate or _onUpgrade, the get database method is returning the database object to the application before it's ready, which then results in "exception: no such table" errors.

If I make a manual call to "get database", and add a 500ms delay before allowing my application to initialize and read any data everything completes fine, but this is clearly not ideal, and from the database helper examples given I don't see any way to avoid this as the issue seems to be fundamentally linked to how openDatabase calls these functions.  I consider myself an amateur coder, so I may be missing something here, but I don't see how this could be working for so many people without errors.

In essence, the "await _initDatabase();" line here does not wait for the database to complete an _onCreate or _onUpgrade action before returning:

```
  Future<Database> get database async {
    if (_database != null) return _database;
    _database = await _initDatabase();
    return _database;
  }
```

In my application, if I add some print statements to show the flow, the initial opening of the database generates exceptions on create or upgrade, as the application flow is something like:

```
Creating database
... table1 created
... table2 created
** _initDatabase completes, and returns database to application **
... table3 created
** Application begins loading data **
... data read from table1
... data read from table2
... data read from table3
... exeption: no such table: table4
... exeption: no such table: table5
... exeption: no such table: table6
... table4 created
... table5 created
... table6 created
```

**Comment by davidmartos96** [comment]
It could be because you are missing an await inside your `initDatabase` method?

Could you share it?

**Comment by myxiplx** [comment]
Here you go, it's essentially unchanged from the example DatabaseHelper:

```
  // open the database (will create or upgrade as required)
  _initDatabase() async {
    // The path_provider plugin gets the right directory for Android or iOS.
    Directory documentsDirectory = await getApplicationDocumentsDirectory();
    String path = join(documentsDirectory.path, _databaseName);
    print("Database path:" + path);
    // Open the database. Can also add an onUpdate callback parameter.
    return await openDatabase(path,
        version: _databaseVersion,
        onCreate: _onCreate,        //If the database doesn't exist, SQflite will call the onCreate function
        onUpgrade: _onUpgrade, //This is called if the database version is higher than before
        onDowngrade: _onDowngrade);
  }
```

**Comment by myxiplx** [comment]
Aaah, got it!  Your question got me thinking down different lines.  It wasn't not enough "await" lines, it was too many "async".  If I remove async from the _onCreate, _onUpgrade, and _onDowngrade functions it works fine.

**Reply by davidmartos96** [reply]
Great that you found it! I would recommend you to install a linter such as `pedantic` or `lint` from pub.dev. They come with predefined lints that help to avoid some common issues and silly mistakes, some of them regarding Futures and async functions.
